plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,553 @@
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 900);
+
+	if (!algoltek_device_wrr(self, 0x80AD, 0, error)) {
+		g_prefix_error(error, "data write failure: ");
+		return FALSE;
+	}
+
+	if (!algoltek_device_wrr(self, 0x80C0, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80C9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80D1, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80D9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80E1, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80E9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_rst(self, 0, error)) {
+		g_prefix_error(error, "system reboot failure: ");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 500);
+
+	/* get ISP image */
+	blob_isp = fu_firmware_get_image_by_id_bytes(firmware, "ISP", error);
+	if (blob_isp == NULL)
+		return FALSE;
+	if (!algoltek_device_ISP(self,
+				 blob_isp,
+				 AG_ISP_ADDR,
+				 fu_progress_get_child(progress),
+				 error)) {
+		g_prefix_error(error, "isp failure: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	if (!algoltek_device_bot(self, AG_ISP_ADDR, error)) {
+		g_prefix_error(error, "system boot failure: ");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 1000);
+
+	if (!algoltek_device_ers(self, error)) {
+		g_prefix_error(error, "data clear failure: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	fu_device_sleep(FU_DEVICE(self), 500);
+
+	/* get payload image */
+	blob_payload = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (blob_payload == NULL)
+		return FALSE;
+	if (!algoltek_device_wrf(self, blob_payload, fu_progress_get_child(progress), error)) {
+		g_prefix_error(error, "data write failure: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	if (!algoltek_device_rst(self, 0x100, error)) {
+		g_prefix_error(error, "system reboot failure: ");
+		return FALSE;
+	}
+	/* success! */
+	return TRUE;
+}
+
+static void
+fu_algoltek_usb_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "reload");
+}
+
+static void
+fu_algoltek_usb_device_init(FuAlgoltekUsbDevice *self)
+{
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_PLAIN);
+	fu_device_add_protocol(FU_DEVICE(self), "tw.com.algoltek.usb");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UNSIGNED_PAYLOAD);
+	fu_device_set_firmware_gtype(FU_DEVICE(self), FU_TYPE_ALGOLTEK_USB_FIRMWARE);
+}
```

Comment:
Since this isn't marked dual bank or self recovery and this is USB what exactly happens if the device is unplugged mid update?

Does it come up in a recovery mode? Can a user still recover?

Reply:
If there is any failure during updating process, the factory default firmware in Mask ROM will be used. User can update the firmware in SPI Flash again to have better video source/sink compatibility by leveraging LVFS or Firmware Upgrade Utility(Windows/macOS/Linux).

Reply:
Thanks